### DISCUSSION TITLE
Ability for Start/Stop Fan Setup without spin up / spin down - loop

### DISCUSSION THREAD

**Discussion by bluespresso** [Rem0o/FanControl.Releases#862] [Ideas]
As for now I do not see how I can accomplish a good setup where the fans are allowed to stop on an idle computer. The main problem is that I cannot set up two different thresholds for fan stop and fan start. The lower temperature threshold for fan stop would have to be several degrees lower than the threshold for fan start. Otherwise there always exist load scenarios where the fan in a loop spins up and stops in short intervals because the difference in cooling between full stop and minimal rpm is enough to overcome this small load.
Two solutions seem possible:
1)	In one graph one could put two different curves, one for fan stop state and one for fan spinning state. Here one could configure lower and upper thresholds graphically.
2)	Add another parameter for the fan: Requested load % for fan to spin up from stop. Then one could set up this parameter several percent higher than stop % and would accomplish the same thing. I initially thought the current parameter start % would do exactly that. But it is something different, the percentage of power the fan is powered with on start up while the threshold for a start up request remains the parameter stop %.

**Comment by Rem0o** [comment]
> while the threshold for a start up request remains the parameter stop %.
That's not how it works.

Fan stop and start are 2 different thresholds.
Start% should be higher than stop%.   Start% will serve as a "kick-start", while stop% is used as a threshold to snap to 0%.
You can run your fan between those 2 values.

**Reply by bluespresso** [reply]
Sure, I understand. But I mean something different. Let´s say you have set:
Start% = 25
Stop%=20
You also have a linear controlling curve for the fan starting below 20% because you want the fan to be completely off when the computer is idle. The curve is rising with temperature.
As it is now the fan would be off with 20% and below but would start spinning as soon as the control curve reaches 21% and for a kick start drive it with 25% for a short time. But…
In order to make a practical solution for a start/stop scenario I would need another option so the fan only starts up when the control curve demands for example 35% or more. The thermal load corresponding to this condition should be enough to keep the temperature above the 20% demand of the control curve.

**Reply by Rem0o** [reply]
> In order to make a practical solution for a start/stop scenario I would need another option so the fan only starts up when the control curve demands for example 35% or more. The thermal load corresponding to this condition should be enough to keep the temperature above the 20% demand of the control curve.

Then simply make a graph that goes from 0 to 35% (almost vertical line) at a specific temperature, then increases smoothly up to whatever % from there.

You can also make asymmetric stuff with the target fan curve, and then mix it with a normal graph to get the linear behavior also. You can also put really high hysteresis to help with situations like this.

If you set your curve in a specific way so that it will loop around the start and stop% continuously, then your curve is most likely badly configured for your scenario in the first place. Your configured "minimum temp" on your curve for the fans to start spinning is most likely way too low. You don't want to put the "start/stop" region of % of a fan right in the "core" of a fan curve.   Myself I do have a 0 rpm setup for my GPU: it fits the thermal state of my GPU so that at idle it will keep 0 rpm (38 to 42 deg typically), and only come out if a significative load ( > 45 deg C ) comes up for more than 5 seconds.

**Comment by liamauve** [comment]
what might help is to set a very small step down %/sec, so even when the curve says it should be at 0rpm now, it will take a few seconds for it to actually stop.

but i think the proper way to do this is with hysteresis, it even has an option to only apply on the way down. but personally i dont quite understand how it works. 

for example if i set the hysteresis to 30c and 1sec, it should mean, according to the pop up description, that the temperature must drop by 30c in the given 1 second response time "in order for the fan curve to change value". however in reality my temperature will never drop by 30c within 1 second, more like 20-25c the first second the load is over. so if that condition is never met, it should never spin down? but irl it does.. ?

**Reply by Rem0o** [reply]
@liamauve hysteresis is ignored going into limits.

**Comment by bluespresso** [comment]
Ok, it is not obvious at all but with your help I managed to make the configuration I desired. It is the asymmetric stuff I needed from target fan curve.
Explanation: 
Control GPU fan dependent on GPU temperature.
GPU fan minimum spin rate is @7%, about 250 rpm.
GPU fan shall not spin when GPU temperature is below 45°C.
GPU fan shall only start spinning a@7%, 250rpm when GPU temperature hits 55°C or above under load.
If the assumed constant load is sufficiently small the temperature will drop below 55°C when the GPU fan spins up from 0% to 7%@ 250 rpm but will not go as low as 45°C.
Only when the load is completely gone and GPU is idle the GPU temperature will drop to 45°C or below with the fan still spinning @7%, 250 rpm. Then the GPU fan stops. 
To accomplish this I need three different curves:
1)	Graph fan curve:  20°C – 55°C = 0%; linear from 55°C = 0% to 70°C = 93%, above 70°C = 93%
2)	Target fan curve: Idle temp = 45°C; idle fan speed = 0%; load temp = 55°C; load fan speed = 7%
3)	Mix fan curve: graph fan curve + target fan curve (note 93% from graph + 7% from target = 100% max fan speed)

![01](https://user-images.githubusercontent.com/58310550/172700255-2cce1390-946d-4bf2-8552-9ed6cc923689.jpg)
![02](https://user-images.githubusercontent.com/58310550/172700269-fab04a91-998f-4bfe-bcae-5e0c98e24d82.jpg)

From this I learn that maybe the interface is not self-explanatory.  Target fan curve might be renamed to something like hysteresis fan curve. One would probably need further explanation of the different curve styles to reach a solution like this without the support forum.

But overall good job on the software so far. I am really happy that a solution like this exists.

**Reply by Rem0o** [reply]
I would use "max" instead of "sum" in this case. 

You could also remove the target fan curve all together and play with the hysteresis values of the graph, like setting 10 degree with "only on the way down" checked.

> From this I learn that maybe the interface is not self-explanatory 

Well, your use case is realllllllly specific for some reason and has asymmetric behavior. Most people, like myself, just set a graph and call it a day, then maybe play with the hysteresis values a bit. Made a UI/UX that is simple enough for common scenarios, but with enough customization options to make really complex behaviors like the one you just did. The balance is tricky between the two.

**Reply by bluespresso** [reply]
I played around with the hysteresis values of graph but it didn´t do anything for me. Could not figure it out. Is there more documentation on this?

**Reply by Rem0o** [reply]
Hysteresis in a single picture : https://en.wikipedia.org/wiki/Hysteresis#/media/File:Hysteresis_sharp_curve.svg

Basically, its the amount of change required for something to happen. A 3 degree hysteresis will require a 3 degree change before the curve actually sets a new %.

**Reply by bluespresso** [reply]
Ok, I know hysteresis. But for me it is not clear how it applies in graph curve. With my approach I have the hysteresis well defined between 45°C and 55°C and 0% fan power and 7% fan power. I don´t see how this could be done with the two parameters in graph curve (only on the way down).
